### PR TITLE
ch4/ucx: enable multiple vci in UCX

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

This PR enables pt2pt multiple vci communication in UCX.

Reference #4530 for OFI counterpart.

[skip warnings]
## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/custom --with-ch4-max-vcis=4 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by hzhou** [comment]
test:mpich/custom --with-ch4-max-vcis=4 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by hzhou** [comment]
test:mpich/custom --with-ch4-max-vcis=4 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by hzhou** [comment]
****test:mpich/custom --with-ch4-max-vcis=4 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by hzhou** [comment]
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by hzhou** [comment]
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by hzhou** [comment]
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by hzhou** [comment]
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4
testlist: threads

**Comment by hzhou** [comment]
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4
testlist: threads/pt2pt

**Comment by hzhou** [comment]
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

test:mpich/ch4/ucx

File: src/mpid/ch4/netmod/ucx/ucx_init.c [hunk 1]
```diff
@@ -229,7 +229,8 @@ int MPIDI_UCX_mpi_init_hook(int rank, int size, int appnum, int *tag_bits, MPIR_
 
     ucp_params.field_mask = UCP_PARAM_FIELD_FEATURES |
         UCP_PARAM_FIELD_REQUEST_SIZE |
-        UCP_PARAM_FIELD_ESTIMATED_NUM_EPS | UCP_PARAM_FIELD_REQUEST_INIT;
+        UCP_PARAM_FIELD_ESTIMATED_NUM_EPS |
+        UCP_PARAM_FIELD_REQUEST_INIT | UCP_PARAM_FIELD_MT_WORKERS_SHARED;
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ucx/ucx_init.c:233]: What implications does that have to overheads in the UCX library? Should `UCP_PARAM_FIELD_MT_WORKERS_SHARED` be conditionally added based on MPICH configuration?

Comment (hzhou) [reply src/mpid/ch4/netmod/ucx/ucx_init.c:233]: Good point, that can be done.

Comment (hzhou) [reply src/mpid/ch4/netmod/ucx/ucx_init.c]: fixed.

**Comment by hzhou** [comment]
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4
✔️
test:mpich/ch4/ucx
✔️

**Comment by hzhou** [comment]
test:mpich/warnings/auto

**Comment by hzhou** [comment]
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by hzhou** [comment]
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Comment by hzhou** [comment]
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

File: src/mpid/ch4/netmod/ucx/ucx_init.c [hunk 2]
```diff
@@ -262,44 +262,55 @@ int MPIDI_UCX_mpi_init_hook(int rank, int size, int appnum, int *tag_bits, MPIR_
...
     comm = MPIR_Process.comm_world;
-    pending = MPL_malloc(sizeof(ucs_status_ptr_t) * comm->local_size, MPL_MEM_OTHER);
-
-    for (i = 0; i < comm->local_size; i++) {
-        ucp_request = ucp_disconnect_nb(MPIDI_UCX_AV(&MPIDIU_get_av(0, i)).dest[0][0]);
-        MPIDI_UCX_CHK_REQUEST(ucp_request);
-        if (ucp_request != UCS_OK) {
-            pending[p] = ucp_request;
-            p++;
+    int n = MPIDI_UCX_global.num_vnis;
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ucx/ucx_init.c:275]: We can squash this commit with the one that initializes all vnis.

File: src/mpid/ch4/netmod/ucx/ucx_win.h [hunk 3]
```diff
@@ -133,7 +159,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_NM_rma_win_cmpl_hook(MPIR_Win * win)
     if (MPIDI_UCX_is_reachable_win(win) && MPIDI_UCX_win_need_flush(win)) {
         ucs_status_t ucp_status;
         /* maybe we just flush all eps here? More efficient for smaller communicators... */
-        ucp_status = ucp_worker_flush(MPIDI_UCX_global.ctx[0].worker);
+        ucp_status = MPIDI_UCX_flush(0);
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ucx/ucx_win.h:162]: Is this change necessary to enable multi-vni pt2pt and collectives?

Comment (hzhou) [reply src/mpid/ch4/netmod/ucx/ucx_win.h:162]: See this comments: https://github.com/pmodels/mpich/pull/4648#issuecomment-651451756

Somehow `ucp_worker_flush` hangs in the test `./rma/rget_unlock 2` unless global progress is polled even though only vci 0 is ever used. I don't quite understand why yet. Replacing `ucp_worker_flush` with the non-blocking version allows polling global progress in it to prevent the deadlock.

Comment (raffenet) [reply src/mpid/ch4/netmod/ucx/ucx_win.h:162]: The old blocking flush function calls this: https://github.com/openucx/ucx/blob/master/src/ucp/rma/rma.inl#L53-L74, which is very similar to `MPIDI_UCX_flush` but only progresses a single worker. I'm concerned that we don't understand what is going on here. I don't think this is ready to be merged. We will talk about this on the call Thursday, but the release is going to be delayed significantly, so this still has time to investigate and get it right.

**Comment by hzhou** [comment]
test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4
✔️
test:mpich/ch4/ucx
✔️

File: src/mpid/ch4/netmod/ucx/ucx_win.h (hunk 3 above)
Comment (hzhou) [reply src/mpid/ch4/netmod/ucx/ucx_win.h:162]: Sounds good.

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

File: src/mpid/ch4/netmod/ucx/ucx_win.h (hunk 3 above)
Comment (hzhou) [reply src/mpid/ch4/netmod/ucx/ucx_win.h:160]: I found the root issue. When ucx creating workers and establishing endpoints (in `all_vnis_address_exchange` ), it is in fact non-blocking operations and it only can be completed/cleared by progress on all workers that are involved. Since we are doing all-to-all wireup, that means we need do progress on all vni workers. UCX internally only maintain a single `flush_ops_count` for each worker. So even though in the test program we only use a single vni and the rma operations are actually cleared, the pending ops on connection to other vni prevents the `ucp_worker_flush` to clear.

The fix is to force flush all workers during `MPIDI_UCX_post_init`.

Comment (hzhou) [reply src/mpid/ch4/netmod/ucx/ucx_win.h:160]: @rzambre Refer to my comment above. In your case since you are only doing one-to-one connections rather than all-to-all connections, you won't have the same issue.

**Comment by hzhou** [comment]
Added commit to flush all workers during init. The commit to "replace ucp_worker_flush with ucp_worker_flush_nb" is now inessential. I left it in since `ucp_worker_flush` is deprecated and it's probably a good idea to replace anyway.

test:mpich/custom --with-ch4-max-vcis=8 --enable-thread-cs=per-vci
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

test:mpich/ch4/ucx

**Comment by hzhou** [comment]
Tests passed. @raffenet Could you take one more look to see if it can make into the current release, since I have addressed the ucp_worker_flush issue?
